content/docs/ai/ai-checkpoints.md

```diff
@@ -0,0 +1,356 @@
+</Admonition>
+
+#### Restore workflow
+
+```json
+{
+  "target_branch_id": "br-active-branch-123456", // Your root branch ID
+  "finalize_restore": true // Moves computes and preserves connection string
+}
+```
+
+Note: This shows a rollback to the active branch. For preview environments, you would omit `target_branch_id` and set `finalize_restore: false`.
+
+**Important:** When restoring with `finalize_restore: true`, your previous active branch becomes a backup branch named `production (old)` or similar. This backup branch preserves your pre-restore state but should be deleted during cleanup to avoid unnecessary costs.
+
+After calling the restore API:
+
+1. Extract operation IDs from the response (may be empty array)
+2. Poll each operation until it reaches a terminal state
+3. **Do not connect yet** - database still points to old state
+4. Only after all operations complete can you connect to the restored snapshot
+
+**Polling operations:**
+
+```javascript
+// Pseudo-code for polling operations
+function waitForOperation(operationId) {
+  while (true) {
+    status = getOperationStatus(operationId)  // GET /projects/{project_id}/operations/{operationId}
+
+    // Terminal states - safe to proceed
+    if (['finished', 'skipped', 'cancelled'].includes(status)) return
+
+    // Error state - handle appropriately
+    if (status === 'failed') throw error
+
+    sleep(5 seconds)
+  }
+}
+
+// After restore API call
+operationIds = restoreResponse.operations.map(op => op.id)
+if (operationIds.length > 0) {
+  operationIds.forEach(id => waitForOperation(id))
+}
+// NOW safe to connect
+```
+
+<Admonition type="tip" title="Working example">
+See `lib/neon/apply-snapshot.ts` and `lib/neon/operations.ts` in the demonstration application at https://github.com/neondatabase-labs/snapshots-as-checkpoints-demo
+</Admonition>
+
+<Admonition type="note" title="todo">
+**Visual Diagram: Operations Polling Flow** - Flowchart showing restore API call, operations returned, polling loop with state transitions (running → finished/failed), danger zones where connection attempts fail, and the final "safe to connect" state
+</Admonition>
+
+**Common restore errors:**
+
+- **Connection to old state**: Ensure all operations completed
+- **Target branch not found**: Verify branch exists
+- **Operation timeout**: Retry with longer timeout
+- **Accumulating backup branches**: Delete old backup branches (e.g., `production (old)`) after successful restore verification
+
+#### Preview environments
+
+Create a temporary branch to preview a checkpoint without affecting the active branch:
+
+```json
+{
+  "name": "preview-checkpoint-123",
+  "finalize_restore": false // Creates new branch for preview without moving computes
+}
+```
+
+This creates a new branch with its own connection string for preview. The active branch remains unchanged.
+
+### Managing checkpoints
+
+#### List available checkpoints
```

Comment:
Should we provide curl examples for each here and for other API calls? We could look at providing SDK examples in tabs  (e.g., TypeScript). Maybe we can ask Andre what thinks would be useful.